### PR TITLE
[RFC] vim-patch:7.4.1126

### PR DESCRIPTION
**Pull request by jamessan**
Problem:    Can only get the directory of the current window.
Solution:   Add window and tab arguments to getcwd() and haslocaldir().
            (Thinca, Hirohito Higashi)

https://github.com/vim/vim/commit/c970330676eaae7ba7cd05cfa46df5a413853ef9

### CODE REVIEW

**Comment by justinmk** [comment]
See also https://github.com/neovim/neovim/pull/4379. Might need @HiPhish to comment on this.

**Comment by jamessan** [comment]
Bah, could've sworn I double-checked there wasn't an open PR before starting work on this.

At any rate, `find_tabwin` isn't useful for neovim's `getcwd`/`haslocaldir` since it has the concept of tab-local working directory, however future Vim patches do make use of `find_tabwin`.

The tests that Bram made for `getcwd`/`haslocaldir` also exposed some problems in the tab-local working directory implementation that I'm currently working on fixing (and should have a PR posted for in a day or two).

**Comment by HiPhish** [comment]
If I read the PR correctly it changes some names in the documentation and the default scope is maximum instead of minimum?
```
CdScope scope = MAX_CD_SCOPE;
```

I don't understand why you would want that behaviour. In Vim if I run `:lcd` in a window and then `getcwd()` (without arguments) I get the scope of the window, i.e. the deepest scope. I wrote the implementation such that without arguments it would work just like it did before. Maybe I'm reading this wrong, what are you trying to accomplish?

**Comment by jamessan** [comment]
It'll make more sense when I separate out the bug fix from the vim-patch merge. :) I pushed this when I realized there were problems with precedence in `getcwd`/`haslocaldir` and wanted to let people know I was working on it to avoid duplication (which I did a poor job checking myself).

**Comment by HiPhish** [comment]
OK. I don't have the time to follow Vim and Neovim development regularly, but if you reference my name I will get a notification if you have a question.

**Comment by jamessan** [comment]
Hmm, so I kept the `find_tabwin()` function that 7.4.1126 added, but it's not currently being used in the code base.  It will be used by future patches, though.  This causes the CI to fail because we have an unused function.

Should I find the next patch that uses that function and add it to this PR?

**Comment by jamessan** [comment]
Need to finish porting the test ...

**Comment by jamessan** [comment]
The QB test hangs are unrelated to this PR.  I'm able to reproduce them with HEAD too when xclip or xsel are in the environment.  test/functional/legacy/eval_spec.lua hangs with either present.  test/functional/shell/viml_system_spec.lua hangs only when xclip is present (since it only enables that testing with xclip).

**Comment by justinmk** [comment]
https://github.com/neovim/neovim/pull/4798 should fix the hangs. Might be worth merging it here to confirm.

**Comment by jamessan** [comment]
Yeah, I tried that and it didn't help.  Of course, now that I've tried bisecting I can't reproduce the hang anymore...

File: test/functional/legacy/getcwd_spec.lua [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+describe('getcwd', function()
+  before_each(clear)
+
+  after_each(function()
+    helpers.rmdir('Xtopdir')
+  end)
+
+  it('is working', function()
+    -- Do all test in a separate window to avoid E211 when we recursively.
+    -- Delete the Xtopdir directory during cleanup.
```
Comment (justinmk) [inline test/functional/legacy/getcwd_spec.lua]: some punctuation quirks here (from the conversion script I guess)

File: test/functional/legacy/getcwd_spec.lua [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+
+    -- This will cause a few errors, do it silently.
```
Comment (justinmk) [inline test/functional/legacy/getcwd_spec.lua]: comment is no longer relevant?

Comment (jamessan) [inline test/functional/legacy/getcwd_spec.lua]: Right.  Removed all those comments.

**Comment by jamessan** [comment]
@jszakmeister Any chance you could help debug the QB failures? The locations of the timeouts seem to be pretty reliable on each builder flavor, yet are distinct from each other.  I can't reproduce the problem manually.  The issues I [thought] were the problem can't happen since they require xsel/xclip to be available (c.f., #4900).

[thought]: https://github.com/neovim/neovim/pull/4841#issuecomment-224156271

**Comment by justinmk** [comment]
If I recall, @jszakmeister mentioned that we weren't doing a good job of cleaning up some sockets that later caused problems because of insufficient permissions to delete them.

**Comment by jamessan** [comment]
So how can we reproduce those issues in other environments?  Having test failures that only exhibit themselves on QB are a pain to diagnose.

**Comment by justinmk** [comment]
Found the discussion I was thinking of: https://github.com/neovim/neovim/pull/4718#issuecomment-217811363 Not sure if that's relevant anymore. 

Unlike QB, travis starts with a wiped machine each time. QB keeps us a bit more honest, since it only cleans up some files but not the entire environment. @jszakmeister it may help to know the exact sequence of commands that QB invokes so that we can run this in local environments to try to reproduce these tests that are tripping over each other. We should commit that script to `.ci/quickbuild.sh`.

**Comment by KillTheMule** [comment]
It's [decode_spec.lua](https://github.com/neovim/neovim/issues/4685) again. You can see it by checking the log for build-and-run-tests (no idea how to link that). At least, that test is throwing that error again, so I assume it's hannging. There's also an error `E94: No matching buffer for _test_` in there, no idea where that belongs. You can search for `WARN` to get those errors.

**Comment by justinmk** [comment]
> `E94: No matching buffer for _test_` ... You can search for WARN to get those errors.

Those WARN lines are just QB's way of reporting stderr. It's unfortunately noisy, but this is a known quirk of the unit tests which are testing for those error messages intentionally. So just ignore that.

**Comment by jamessan** [comment]
> It's decode_spec.lua again.

Indeed it is.  I thought those issues (which I initially saw when porting 7.4.613 in #4630) were fixed by @ZyX-I in #4852.  

It'd be cool if at least the Linux QB ran with `ulimit -c unlimited` and if a core is found after a test phase the backtrace were logged.

**Comment by justinmk** [comment]
> It'd be cool if at least the Linux QB ran with ulimit -c unlimited and if a core is found after a test phase the backtrace were logged.

We do attempt that here: https://github.com/neovim/neovim/blob/e7ab3e7e591c739/.ci/common/test.sh#L1             

But where are you seeing a core dump? I only saw SIGTERM, which was sent by QB itself because of the 20 minute timeout.

**Comment by jamessan** [comment]
> We do attempt that here: https://github.com/neovim/neovim/blob/e7ab3e7e591c739/.ci/common/test.sh#L1

That's Travis, not QB.

> But where are you seeing a core dump?

```
17:08:29,643 INFO  - # test/unit/buffer_spec.lua @ 293: buffer functions build_stl_str_hl should print the number of lines in the buffer
17:08:29,643 INFO  - ok 19 - buffer functions build_stl_str_hl should print the number of lines in the buffer
17:08:29,735 WARN  - E94: No matching buffer for test
17:08:29,736 WARN  - E474: No container to close: ]
17:08:29,737 WARN  - E474: Closing list with curly bracket: }
```

That's the same behavior that was seen in #4630 when the `line_breakcheck` call caused a crash in the buffer_spec unittest.

**Comment by justinmk** [comment]
Running `TEST_FILE=test/unit/buffer_spec.lua make unittest` with [this line](https://github.com/neovim/neovim/blob/e7ab3e7e591c739/cmake/RunTests.cmake#L33) removed shows only the `E94: No matching buffer for _test_` messages.

So you're right, these are unintended errors.

**Comment by justinmk** [comment]
Running `make unittest` with [this line](https://github.com/neovim/neovim/blob/e7ab3e7e591c739/cmake/RunTests.cmake#L33) removed shows these Vim user-land errors:

```
E474: No container to close: ]                                        
E474: Closing list with curly bracket: }                                                                         
E474: Closing dictionary with square bracket: ]                                                                  
E474: Trailing comma: ]                                                                                          
E474: Expected value after colon: }                                                                              
E474: Expected value: }                                                                                          
E474: Comma not inside container: ,                                                                              
E474: Duplicate comma: ,1]                                                                                       
E474: Comma after colon: ,}                                                                                      
E474: Using comma in place of colon: ,}                                                                          
E474: Leading comma: ,}                                                                                          
E474: Leading comma: ,]                                                                                          
E474: Colon not inside container: :                                                                              
E474: Using colon not in dictionary: :]                                                                          
E474: Unexpected colon: :}                                                                                       
E474: Duplicate colon: :1}                                                                                       
E474: Expected null: n                                                                                           
E474: Expected true: t                                                                                           
E474: Expected false: f                                                                                          
E474: Unfinished escape sequence: "\                                                                             
E474: Unfinished unicode escape sequence: "\u"                                                                   
E474: Expected four hex digits after \u: \uXXXX"                                                                 
E474: Unknown escape sequence: \?"                                                                               
E474: ASCII control characters cannot be present inside string: ^I"                                              
E474: Only UTF-8 strings allowed: <c2>"                                                                          
E474: Only UTF-8 code points up to U+10FFFF are allowed to appear unescaped: ??                                  
E474: Expected string end: "                                                                                     
E474: Failed to convert string "ꯍ" from UTF-8                                                                    
E474: Missing number after minus sign: -         
E474: Missing number after decimal dot: -1.                                                                      
E474: Missing exponent: -1.0e                                                                                    
E474: Unidentified byte: ?                                                                                       
E474: Trailing characters: ?                                                                                     
E474: Unexpected end of input: [1●                                                                               
E474: Missing number after minus sign: -●                                                                        
●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●E94: No matching buffer for _test_                               
E94: No matching buffer for _test_                                                                               
E94: No matching buffer for _test_●                                                                              
E94: No matching buffer for test
```

These are errors that are intentionally tickled by the unit tests. The fact that they appeared in https://github.com/neovim/neovim/pull/4630 is a red herring, because they _always_ appear when the build fails _just in case_ they actually contain something unexpected.

**Comment by jamessan** [comment]
Ah, ok.  So we still have no idea what's causing the problem. :)

**Comment by justinmk** [comment]
💯 👍

**Pull request merged**